infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/AddNodeWorkflow.java

```diff
@@ -0,0 +1,109 @@
+package org.corfudb.infrastructure.orchestrator;
+
+import org.corfudb.protocols.wireprotocol.orchestrator.AddNodeRequest;
+import org.corfudb.protocols.wireprotocol.orchestrator.Request;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Layout;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import static org.corfudb.format.Types.OrchestratorRequestType.ADD_NODE;
+
+/**
+ * A definition of a workflow that adds a new node to the cluster.
+ *
+ * @author Maithem
+ */
+@NotThreadSafe
+public class AddNodeWorkflow implements Workflow {
+
+    final AddNodeRequest request;
+
+    public AddNodeWorkflow(Request request) {
+        this.request = (AddNodeRequest) request;
+    }
+
+    @Override
+    public String getName() {
+        return ADD_NODE.toString();
+    }
+
+    @Override
+    public List<Action> getActions() {
+        return null;
+    }
+
+    class BootstrapNode implements Action {
+        @Override
+        public String getName() {
+            return "BootstrapNode";
+        }
+
+        @Override
+        public Map<String, Object> execute(@Nonnull CorfuRuntime runtime) throws Exception {
+            runtime.getLayoutManagementView().bootstrapNewNode(request.getEndpoint());
+            return null;
+        }
+    }
+
+    /**
+     * This action adds a new node to the layout. If it is also
+     * added as a logunit server, then in addition to adding
+     * the node the address space segment is split at the
+     * tail determined during the layout modification.
+     */
+    class AddNodeToLayout implements Action {
+        @Override
+        public String getName() {
+            return "AddNodeToLayout";
+        }
+
+        @Override
+        public Map<String, Object> execute(@Nonnull CorfuRuntime runtime) throws Exception {
+            Layout currentLayout = runtime.getLayoutView().getLayout();
```

Comment:
You will need to clone this layout before you pass it to the addNode method.
The add node method tries to seal this layout which will modify it.

Reply:
There's a race between the invalidate layout and clone.

Reply:
Actually `runtime.getLayoutView().getLayout()` should block on the future and wait for the runtime to return the layout after the `invalidateLayout`.
But you are right, there is some race here, I have seen it too but have not been able to spot the exact reason.